packages/back-end/src/queryRunners/QueryRunner.ts

```diff
@@ -656,29 +652,31 @@ export abstract class QueryRunner<
     logger.debug("Creating query for: " + name);
     const concurrencyLimitReached = await this.concurrencyLimitReached();
     const dependenciesComplete = dependencies.length === 0;
-    const readyToRun =
-      dependenciesComplete && !runAtEnd && !concurrencyLimitReached;
+    const readyToRun = dependenciesComplete && !runAtEnd;
     const doc = await createNewQuery({
       query,
       queryType,
+      dependencies,
+      runAtEnd,
       datasource: this.integration.datasource.id,
       organization: this.integration.context.org.id,
       language: this.integration.getSourceProperties().queryLanguage,
-      dependencies: dependencies,
-      running: readyToRun,
-      runAtEnd: runAtEnd,
+      running: readyToRun && !concurrencyLimitReached,
     });
 
+    // Update heartbeat for the query once every 30 seconds
+    // This lets us detect orphaned queries where the thread died
+    this.updateHeartbeatOnInterval(doc);
```

Comment:
This happens before the queries are split to be handled separately based on queued/running status, so now all queries will have a heartbeat